src/StringHelper.php

```diff
@@ -580,8 +580,13 @@ public static function parsePath(
      */
     public static function trim(string|array $string, string $pattern = self::DEFAULT_WHITESPACE_PATTERN): string|array
     {
+        self::ensureUtf8String($string);
         self::ensureUtf8Pattern($pattern);
 
+        /**
+         * @var string|string[] `$string` is correct UTF-8 string and `$pattern` is correct (it should be passed
+         * already prepared), so `preg_replace` never returns `null`.
+         */
         return preg_replace("#^[$pattern]+|[$pattern]+$#uD", '', $string);
```

Comment:
Maybe add a private method
```php
private static function trimUsingPregReplace(string|array $string, string $pattern): string|array
    return preg_replace($pattern, '', $string) ?? throw new InvalidArgumentException('String is not a valid UTF-8 string.');
}
```

and call the method as follow `return $this->trimUsingPregReplace("#^[$pattern]+|[$pattern]+$#uD", $string);`

This will avoid extra `foreach` and `preg_match('##u', $s)`

Reply:
When `$string` is array of strings, invalid UTF-8 string just ignored in `preg_replace()`.